### PR TITLE
Add hotkeys' descriptions

### PR DESCRIPTION
**Pull request by actionless**
follow-up on #151 

how it looks (atm i added descriptions only for 7 hotkeys):
2015-08-20-214657_3840x2160_scrot

how it looks when vim is opened:
2015-08-20-214754_3840x2160_scrot

screen with more hotkeys (from my own config):
2015-08-20-220442_3840x2160_scrot

some work on appearance is still needed, like adding margins and so on, feel free to suggest

### CODE REVIEW

File: lib/hot/keys/vim.lua [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      ["'"]=". goto mk. BOL",
+      b="prev word",
+      n="next word",
+      [',']="reverse t/T/f/F",
+      ['/']=". find",
+    }
+  },{
+    modifiers = {"Shift",},
+    keys = {
+      ['`']="toggle case",
```
Comment (blueyed) [inline lib/hot/keys/vim.lua]: Is that specific to your keyboard layout?  (Vim uses `~`, which is `AltGr++` on a German layout)

**Comment by blueyed** [comment]
Awesome!

I think that https://github.com/awesomeWM/awesome/pull/365 might help here.  It should allow to use the utf8 representation of keys in the bindings.

File: lib/hot/keys/vim.lua (hunk 1 above)
Comment (actionless) [inline lib/hot/keys/vim.lua]: i am using en_US

**Comment by actionless** [comment]
seems like only the new name and pagination remaining?

**Comment by blueyed** [comment]
What do you think about getting in #365 for this first?

**Comment by actionless** [comment]
from what i understood after #365 will be merged i will just need to remove some predefined labels, but seems like no other changes need to be done in hotkey widget, or i am missing something?

File: lib/hot/widget.lua [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  keygrabber = keygrabber,
+}
+local awful = require("awful")
+local common = require("awful.widget.common")
+local wibox = require("wibox")
+local beautiful = require("beautiful")
+local dpi = beautiful.xresources.apply_dpi
+
+
+local markup = {}
```
Comment (actionless) [inline lib/hot/widget.lua]: btw, what do you think about copying whole `lain.util.markup`[1] into awesome?

[1] https://github.com/copycat-killer/lain/blob/master/util/markup.lua

**Comment by actionless** [comment]
i hope now it should be more readable; do i need to do more visual adjustments?
2015-08-22-123919_3840x2160_scrot

**Comment by actionless** [comment]
i wish someone will make a PR with `mpv` keys later on

**Comment by actionless** [comment]
seems only the last issue remains: https://github.com/awesomeWM/awesome/pull/421#commitcomment-12829419

**Comment by actionless** [comment]
i hope i am done now

**Comment by actionless** [comment]
can you please merge it until it rot?

**Comment by actionless** [comment]
rebased

**Comment by blueyed** [comment]
I'll try this out later.
However, it has a lot of changed to the default config, which makes it hard to adopt/merge into a customized one.

**Comment by actionless** [comment]
it shouldn't break any custom config

**Comment by psychon** [comment]
I haven't tested this and I don't really get the vim-related module, but fine with me. Looking forward to breaking this via per-screen-DPI-textboxes and also looking forward to someone making this work with "non default" keyboard layouts.

**Comment by blueyed** [comment]
> it shouldn't break any custom config

I see.  But you don't get the benefits without merging.

> I don't really get the vim-related module

Me neither.  It will only reflect the default Vim config, but seems to be merely used to provide an example, and therefore appears to be fine as-is.

**Comment by Elv13** [comment]
May I suggest breaking this into 2 PR? One for the `awful.key` changes and one for the widget stuff. I think the `awful.key` changes are quite uncontroversial at this point and could be merged. I think, the remaining part may need some more work. For example, the key id being hardcoded for F keys wont work for all keyboard layouts and so on. I know this commit has been squashed and I am asking to undo some of that, but I think this PR is just too large not to scare off reviewer :P

**Comment by actionless** [comment]
created https://github.com/awesomeWM/awesome/pull/491

**Comment by actionless** [comment]
rebased on top of just-merged #491

**Comment by Elv13** [comment]
The important thing is, Awesome 3.6 will have 22 more lines of useful code sure to be available and well tested by the time it is released ;) (this keybinding story has been going on for a year already, time to start merging some of it, right? :) )

**Comment by actionless** [comment]
mb let's now merge only changes to key descriptions in rc.lua?

**Comment by Elv13** [comment]
Sure

Should the first letter of each comment have a upper case or lower case?

**Comment by actionless** [comment]
i dunno about the case

**Comment by actionless** [comment]
rebased on top of https://github.com/awesomeWM/awesome/pull/519

**Comment by actionless** [comment]
rebased on top of https://github.com/awesomeWM/awesome/pull/542

**Comment by actionless** [comment]
with which application i should replace the vim example?

**Comment by psychon** [comment]
If this should have an application example, I'm fine with vim. It's as good as everything else (and perhaps a bit better, because it's more keyboard-controlled than e.g. gedit).

**Comment by blueyed** [comment]
I agree, Vim is fine.

**Comment by Elv13** [comment]
After all, why not Vim

**Comment by blueyed** [comment]
The default mapping to show the help does not work on a German keyboard layout, where "/" is Shift-7: it will switch to tag 7 instead.

That's probably an issue with the numeric handling of the tag mappings?!


> awful.key({ modkey,           }, "/",      hotkeys_popup.show_help,
>           {description="show help", group="awesome"}),

**Comment by actionless** [comment]
do you have any idea what would be better to do then?

**Comment by psychon** [comment]
Can we use `Mod+?`? Is awesome smart enough to figure out that this is `Mod+Shift+ß` on a German keyboard and `<something else involving Shift>` on an US layout?

**Comment by actionless** [comment]
yup, it's ok i think, if anybody wouldn't like to press 2 modifiers for that than it's not a problem to rebind it later on in the config

**Comment by actionless** [comment]
updated

**Comment by blueyed** [comment]
I use `mod+s` for sticky, but rarely use it.
I think we should figure out how to use `mod+?`.

I also just got that error when trying to call `mod+?` (which is mod-shift-ß) on a de layout) via the following:

    awful.key({ modkey, 'Shift'           }, "ß",      hotkeys_popup.show_help,
                {description="show help", group="awesome"}),
```
W: awesome: luaA_dofunction:77: error while running function
stack traceback:
        /usr/share/awesome/lib/menubar/utils.lua:228: in function 'menubar.utils.compute_textbox_width'
        /usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:251: in local 'insert_keys'
        /usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:264: in upvalue 'create_wibox'
        /usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:362: in function 'awful.hotkeys_popup.widget.show_help'
        /usr/share/awesome/lib/awful/key.lua:57: in function </usr/share/awesome/lib/awful/key.lua:57>
error: /usr/share/awesome/lib/menubar/utils.lua:228: attempt to call a nil value (method 'get_preferred_size')
```
Source: https://github.com/actionless/awesome/blob/6500708225f365f4b13492d1bee3d0e187d3e8c5/lib/awful/hotkeys_popup/widget.lua#L205, called from https://github.com/actionless/awesome/blob/6500708225f365f4b13492d1bee3d0e187d3e8c5/lib/awful/hotkeys_popup/widget.lua#L251-L252.

**Comment by actionless** [comment]
but it works ok for you when it's binded to an other key, right?

i can't reproduce it though:
2015-11-21-224002_3840x2160_scrot

**Comment by blueyed** [comment]
@actionless 
It prints:
```
<span font="Monospace Bold 9"><span foreground="#555555">Shift+Super+</span>ß </span><span font="Monospace 8">show help</span>
```

This apparently comes from:
```
    awful.key({ modkey, 'Shift'           }, "ß",      hotkeys_popup.show_help,
              {description="show help", group="awesome"}),
```

It's really odd though.  If I set it to "foo" it still fails.

```
max_label_content = "foo"
print("wibox", tostring(wibox.widget.textbox(max_label_content)))
```
results in `wibox	wibox.widget.textbox (table: 0x1874ae0)` though.

**Comment by actionless** [comment]
weird, mb it depends on your current locale?

can you try now adding to your config:
```
max_label_content = '<span font="Monospace Bold 9"><span foreground="#555555">Shift+Super+</span>ß </span><span font="Monospace 8">show help</span>'
print("wibox", tostring(wibox.widget.textbox(max_label_content)))
```

**Comment by blueyed** [comment]
I've tried it in awesome-client where it worked, also in the config.

Rebasing on master did not fix it either, but after adding some prints (and removing them again), it now does not happen anymore?!

**Comment by actionless** [comment]
so, is there anything i can improve in this PR?

**Comment by blueyed** [comment]
>  I think we should figure out how to use mod+?.

**Comment by actionless** [comment]
you've told what it could be resolved by https://github.com/awesomeWM/awesome/pull/365

anyway i don't think what handling unicode symbols as keybindings is in scope of this PR and it is not blocking its work (and widget can even display those keybindings not looking to the fact what they are not actually work)

**Comment by blueyed** [comment]
#365 is in since a while, but does not help in this regard.

Anyway, you're right that this should not really block it, and can be changed later anyway.

**Comment by actionless** [comment]
yes, my point was what after that will be fixed, hotkey widget won't require any modifications to display them

**Comment by blueyed** [comment]
Merged finally!
Thanks!

**Comment by actionless** [comment]
thank you guys for reviewing that amount of code :)

**Comment by blueyed** [comment]
@actionless 
Thanks for staying with us! :octocat: :)

**Comment by psychon** [comment]
I just tried this with the default config plus the sky theme:
```
W: awesome: luaA_dofunction:77: error while running function
stack traceback:
	[C]: in function 'error'
	...chon/projects/awesome/build/lib/wibox/widget/textbox.lua:128: in function 'set_markup'
	...chon/projects/awesome/build/lib/wibox/widget/textbox.lua:238: in function <...chon/projects/awesome/build/lib/wibox/widget/textbox.lua:215>
	(...tail calls...)
	...rojects/awesome/build/lib/awful/hotkeys_popup/widget.lua:169: in function 'group_label'
	...rojects/awesome/build/lib/awful/hotkeys_popup/widget.lua:234: in function 'create_wibox'
	...rojects/awesome/build/lib/awful/hotkeys_popup/widget.lua:362: in function 'press'
	/home/psychon/projects/awesome/build/lib/awful/key.lua:57: in function </home/psychon/projects/awesome/build/lib/awful/key.lua:57>
error: ...chon/projects/awesome/build/lib/wibox/widget/textbox.lua:128: Value of 'background' attribute on <span> tag on line 1 could not be parsed; should be a color specification, not 'rgb:80/32/32'
```
No idea where `rgb:80/32/32` comes from. `grep -r rgb:` doesn't find anything.

**Comment by psychon** [comment]
Ohhhh...
```
$ cat .Xdefaults 
[....]
!black
*color0:         rgb:19/19/19
*color8:         rgb:25/25/25
!red
*color1:         rgb:80/32/32
*color9:         rgb:98/2b/2b
!green
*color2:         rgb:5b/76/2f
*color10:        rgb:89/b8/3f
!brown/yellow
*color3:         rgb:aa/99/43
*color11:        rgb:ef/ef/60
!blue
*color4:         rgb:32/4c/80
*color12:        rgb:2b/4f/98
!magenta
*color5:         rgb:70/6c/9a
*color13:        rgb:82/6a/b1
!cyan
*color6:         rgb:92/b1/9e
*color14:        rgb:a1/cd/cd
!white
*color7:         rgb:ff/ff/ff
*color15:        rgb:dd/dd/dd
```
This is a bug in `beautiful.xresources`, right? (It should convert the stuff it reads into something that awesome can understand)

**Comment by blueyed** [comment]
@actionless 
I've tried to use the hotkey_popup for a subset of hotkeys (after overwriting root keys temporarily), but the widget would display all hotkeys.. there is some caching involved, and the widget itself appears to be a singleton(?!) - there's no constructor exposed.
I would like to use a separate instance for the time when the modal mode is enabled.

I've also looked into invoking `show_help` with only a specific (list of) group(s).

Let me know if we should create a separate issue for this.

**Comment by actionless** [comment]
@blueyed 

i can change it to the way what it will be returning popup instance when calling some `.create()` method but it will be returning sorta default instance of the widget when calling it the way how it's called now

are you ok with that idea? it should address your feature-request

___________________

btw, currently i have fork of `menubar` with the same feature implemented (being able to have multiple `menubar` instances instead of only singleton how is it now):
https://github.com/actionless/awesome_config/blob/devel/actionless/menubar/init.lua#L400-L412
but it's too much far from the upstream, do you want to have it upstream?

**Comment by blueyed** [comment]
@actionless 
> i can change it to the way what it will be returning popup instance when calling some .create() method but it will be returning sorta default instance of the widget when calling it the way how it's called now

> are you ok with that idea? it should address your feature-request

Yes! :+1: 

Regarding the menubar, I've created some fix for the current version: https://github.com/awesomeWM/awesome/pull/711.  Having a better implementation by default would be nice, of course.

**Comment by actionless** [comment]
> Regarding the menubar, I've created some fix for the current version: #711. Having a better implementation by default would be nice, of course.

so the point what it's missing at least 4-5 of the latest menubar-related PRs so it could take some time to rebase my menubar fork on upstream, the question is do anyone wants that or not

**Comment by blueyed** [comment]
@actionless 
I want it in general, but have not looked into details.
I am using the menubar only from time to time.

**Comment by actionless** [comment]
i just realized what your solution will be easier (to add an optional argument like `enabled_groups`), no reason for having multiple instances

or it still could be useful for cases like having multiple popups with different hotkeys at the same time?

**Comment by blueyed** [comment]
@actionless 
> or it still could be useful for cases like having multiple popups with different hotkeys at the same time?

Not sure.
Separate instances seem to be better, and do not limit the different choices to available groups.

In my case I would like to have the global instance, and then one for a modal mode and/or some keygrabber (have not checked how useful it's in that context, given that you typically won't use `awful.key` there).

**Comment by actionless** [comment]
created: https://github.com/awesomeWM/awesome/pull/772

**Pull request closed without merging**